srtcore/api.cpp

```diff
@@ -1460,17 +1454,53 @@ int CUDTUnited::groupConnect(CUDTGroup* pg, SRT_SOCKGROUPCONFIG* targets, int ar
             ScopedLock cs(m_GlobControlLock);
             if (m_Sockets.count(sid) == 0)
             {
+                HLOGC(aclog.Debug, log << "srt_connect_group: socket @" << sid << " deleted in process");
                 // Someone deleted the socket in the meantime?
                 // Unlikely, but possible in theory.
                 // Don't delete anyhting - it's alreay done.
                 continue;
             }
-            CUDTGroup::gli_t f = g.add(data);
-            ns->m_IncludedIter = f;
-            ns->m_IncludedGroup = &g;
-            f->weight = targets[tii].weight;
+
+            // There's nothing wrong with preparing the data first
+            // even if this happens for nothing. But now, under the lock
+            // and after checking that the socket still exists, check now
+            // if this succeeded, and then also if the group is still usable.
+            // The group will surely exist because it's set busy, until the
+            // end of this function. But it might be simultaneously requested closed.
+            bool proceed = true;
+
+            if (targets[tii].errorcode != SRT_SUCCESS)
+            {
+                HLOGC(aclog.Debug, log << "srt_connect_group: not processing @" << sid << " due to error in setting options");
+                proceed = false;
+            }
+
+            if (g.m_bClosing)
+            {
+                HLOGC(aclog.Debug, log << "srt_connect_group: not processing @" << sid << " due to CLOSED GROUP $" << g.m_GroupID);
+                proceed = false;
+            }
+
+            if (proceed)
+            {
+                CUDTGroup::gli_t f = g.add(data);
```

Comment:
The group may still change in between cause there is no lock of `m_GroupLock` here. Just the chances are lower now, as those two operations stand close together:
- `if (g.m_bClosing)`
- `g.add(data)`

Reply:
The `m_bClosing` flag is set exclusively in `CUDTGroup::close` function, and it's under `CUDTUnited::m_GlobControlLock`, just like this section. So, the `CUDTGroup::close` if is attempted to be executed at the moment when this section is executed, it will have to wait until this socket is added to the group and the section left. It might then happen that THIS socket will be added to the group, but the NEXT ONE will find out that this group is being closed, but then the whole loop will be interrupted.